pygeoapi/process/routes.py

```diff
@@ -0,0 +1,601 @@
+        for waypoint in way_coords:
+            waypoint_sequence += 1
+            if orig_coordinates is None:
+                orig_coordinates, orig_node_id, dest_node_id = waypoint, None, None
+            else:
+                dest_coordinates = waypoint
+
+                if orig_node_id is None:
+                    orig_node = self.fetch_node_nearby(orig_coordinates)
+                    if orig_node is None:
+                        print("Orig node not found")
+                        return [], None, way_coords[-1]
+                    else:
+                        print("Orig node:", orig_node)
+                        [ orig_node_id, orig_lon, orig_lat, d ] = orig_node
+                dest_node = self.fetch_node_nearby(dest_coordinates)
+                if dest_node is None:
+                    print("Dest node not found")
+                    return [], way_coords[0], None
+                else:
+                    print("Dest node:", dest_node)
+                    [ dest_node_id, dest_lon, dest_lat, d ] = dest_node
+
+                join_query = self.create_join_query(route_def)
+                cost_query = self.create_cost_query(route_def)
+                height_query = self.create_height_query(route_def)
+                weight_query = self.create_weight_query(route_def)
+                obstacles_query = self.create_obstacles_query(route_def)
+                optimization_query = self.create_optimization_query( \
+                    orig_coordinates, dest_coordinates)
+
+                with pgRoutingConnection(self.engine_conn, "",
+                     context="routes") as db:
+                    cursor = db.conn.cursor()
+                    cursor.execute("""SELECT * FROM pgr_bdDijkstra(
+                                      'SELECT w.{0} AS id, w.source, w.target, {1}
+                                      FROM ways AS w {2} WHERE {3} {4} {5} {6}',
+                                      {7}, {8}, directed := true);""".format(
+                                      self.ways_id, cost_query, join_query, 
+                                      height_query, weight_query,
+                                      obstacles_query, optimization_query,
+                                      orig_node_id, dest_node_id))
+                    route_segment_sql = cursor.fetchall()
+
+                # Avoid duplicating intermediate waypoints in sequence
+                if waypoint_sequence < len(way_coords): route_segment_sql.pop()
+                route_seq += route_segment_sql
+
+                orig_coordinates = dest_coordinates
+                orig_node_id, dest_coordinates = dest_node_id, None
+
+        # Return the real coordinates used in route generation
+        if len(route_seq) > 0:
+            return route_seq, [ float(orig_lon), float(orig_lat) ], \
+                [ float(dest_lon), float(dest_lat) ]
+        else:
+            print("Nodes ok but no route found")
```

Comment:
Right, I used these for debugging only.

However, I have a question on these: if the request is valid, but the server is not able to process the request (e.g., the server has insufficient route network data for the  request), it needs to return a response with status code 422, ideally with a meaningful message to guide the user on where the problem is (there is no route from A to B, or startpoint could not be found or so on).

I can implement it in the API Routes part (handling it in api.py) but not in API Processes without modifying the existing code. Currently if there is an execution error in the process (ProcessorExecuteError), API Processes returns with status code 500; otherwise it will be 200 or 201. Is there a way for each process to define its own status code?

Reply:
At first glance, I think you can define a 422 type exception in `pygeoapi.process.base` that your route can use, which then the manager can try/except on and return the appropriate status code.  Then in https://github.com/geopython/pygeoapi/blob/master/pygeoapi/api.py#L2811 you can catch the exception and return the 422.

Reply:
I could create new exceptions and use them in the CRUD functions, which are called directly. However, when calling a process with `self.manager.execute_process` the exception is handled by the manager and little information gets to api script...